### ISSUE TITLE
IAM controller supports AttachRolePolicy

### ISSUE THREAD

**Issue by Julian-Chu** [aws-controllers-k8s/community#1579]
**Is your feature request related to a problem?**
we have some cases to create roles before creating policies, then we need a proper/flexible way to attach policies to roles instead of updating role when new policy created.

**Describe the solution you'd like**
support k8s obj like 
- [IAM API - AttachRolePolicy](https://docs.aws.amazon.com/IAM/latest/APIReference/API_AttachRolePolicy.html)
- [AWS::IAM::ManagedPolicy.Properties.Roles](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-iam-managedpolicy.html)
```yaml
Type: AWS::IAM::ManagedPolicy
Properties: 
  Description: String
  Groups: 
    - String
  ManagedPolicyName: String
  Path: String
  PolicyDocument: Json
  Roles:   # this
    - String
  Users: 
    - String

```
- [crossplane IAMRolePolicyAttachment](https://doc.crds.dev/github.com/crossplane/provider-aws/identity.aws.crossplane.io/IAMRolePolicyAttachment/v1beta1@v0.19.0)

example:
```yaml
kind: IAMRolePolicyAttachment
metadata:
  name: examplepolicyattach
spec:
    roles: 
        - examplerole
    policy: examplepolicy

```

**Comment by jaypipes** [comment]
@Julian-Chu we already support this. A `Role` resource has a `Spec.Policies` field that contains the Policy ARNs to associate with the role. You can also use the `Spec.PolicyRefs` field to refer to one or more `Policy` custom resources that you may have created (or intend to create).

**Comment by Julian-Chu** [comment]
@jaypipes  Thanks for hints.  it will be good alternative solution if we can't have policy attachment API.
Let me explain more detail. We are looking for loose coupling solution no need to update Role yaml when we deploy new resource with  policy.
for example:
```yaml
kind: IAMRolePolicyAttachment
metadata:
  name: examplepolicyattach
spec:
    roles: 
        - examplerole
    policy: examplepolicy
```
We are investigating ack to replace Crossplane.  Crossplane provides this similar policy attachment solution,  and we have something reply on this concept, and think it's not bad to have this flexibility. 😄

**Comment by Julian-Chu** [comment]
@jaypipes   updating the `Policy`  is also a good solution for us 🍺

**Comment by jaypipes** [comment]
@Julian-Chu OK, an update on this...

I spent a few hours yesterday coding up support for `Policy.Roles`, `Policy.Users` and `Policy.Groups`. I got most of the way through implementing it and then realized there was a big problem with implementing these many-to-many relationships from BOTH sides of the relation, as opposed to only a single side of the relation as is currently implemented (with `Role.Policies`, `User.Policies` and `Group.Policies`).

If I implemented support for `Policy.Roles`, I would need to look up all of the referenced `Role` CRs and ensure that each `Role.Spec.Policies` field contained matching Policy ARNs. Likewise, if a user modified `Role.Policies` to include a customer-defined policy (as opposed to an AWS-managed policy), I would need to look up the `Policy` CR and ensure that its `Policy.Spec.Roles` field contained that matching Policy ARN. Keeping these things in sync would be a bit of a nightmare and would go against GitOps best practices of only having the Kubernetes user define desired state and not a controller (controllers should merely reconcile changes in desired state, not change the desired state itself).

All that said, I'm afraid I'm going to conclude that we will not support a `Policy.Spec.Roles` field and instead only support the existing mechanism of using `Role.Spec.Policies` to assign policies to a role. Same goes for User and Group policies. I think that this user experience (of ensuring that the Role, User and Group's `Policies` field is the single source of desired state truth for policy-to-entity relationships is actually the most user-friendly way of describing these relationships and since the whole `Policy.Spec.Roles` field was just an additional way to express the same relationship, it's OK to not add this functionality into the IAM controller.

Hope the above explanation makes sense!
-jay
